README.md
[![Clojars Project](https://img.shields.io/clojars/v/org.clj-commons/byte-transforms.svg)](https://clojars.org/org.clj-commons/byte-transforms)
[![cljdoc badge](https://cljdoc.org/badge/org.clj-commons/byte-transforms)](https://cljdoc.org/d/clj-commons/byte-transforms)
[![CircleCI](https://circleci.com/gh/clj-commons/byte-transforms.svg?style=svg)](https://circleci.com/gh/clj-commons/byte-transforms)

This library helps you **hash**, **compress**, and **encode** streams of bytes.  It contains the methods in the standard Java lib, as well as a curated collection of the best available methods.

## Usage

#### Leiningen
```clojure
[org.clj-commons/byte-transforms "0.2.1"]
```
#### deps.edn
```clojure
org.clj-commons/byte-transforms {:mvn/version "0.2.1"}
```

All functions are in the `byte-transforms` namespace.  There are five primary functions, `hash`, `compress`, `decompress`, `encode`, and `decode`.  Each takes three arguments: the bytes, the method, and an (optional) options map.  The bytes can be anything which is part of the [byte-stream](https://github.com/ztellman/byte-streams) conversion graph.

```clojure
byte-transforms> (hash "hello" :murmur64)
2191231550387646743

byte-transforms> (compress "hello" :snappy)
#<byte[] [B@7e0f980b>
byte-transforms> (byte-streams/to-string (decompress *1 :snappy))
"hello"

byte-transforms> (byte-streams/to-string (encode "hello" :base64 {:url-safe? false}))
"aGVsbG8"
byte-transforms> (byte-streams/to-string (decode *1 :base64))
"hello"
```

Note that Base64 encoding defaults to the [URL-safe](https://en.wikipedia.org/wiki/Base64#URL_applications) variant, which means that the output will not be padded.  This can be disabled by passing in `{:url-safe? false}` to `encode`.

Available methods can be found via `available-hash-functions`, `available-compressors`, and `available-encoders`:

```clojure
byte-transforms> (available-hash-functions)
(:sha384 :md2 :crc32 :crc64 :sha512 :sha1 :murmur32 :murmur128 :adler32 :sha256 :md5 :murmur64)

byte-transforms> (available-compressors)
(:lz4 :bzip2 :snappy :gzip)

byte-transforms> (available-encoders)
(:base64)
```

When choosing a compression algorithm, `snappy` is typically the fastest, `bzip2` yields the highest compression, and `lz4` provides a good balance between higher compression rate and fast decompression.  All the compression algorithms except `lz4` are concat-able; multiple compressed segments can be concatenated and decompressed as a single stream.

Full stats on all methods can be found by cloning the project and running `lein test :benchmark`.

## License

Copyright © 2013 Zachary Tellman

Distributed under the Apache License 2.0.


CHANGELOG.adoc
=== 0.2.1

* Changed group to org.clj-commons

=== 0.2.0

* Bumped snappy-java dep to 1.1.8.3 to support Apple Silicon

Contributions by Zac Xu

=== Pre-0.2.0

Created by Zach Tellman.

Contributions Zach Tellman, Erik Assum, Matthew Davidson, Ganesh Gautam, and Reid Draper.



test/byte_transforms_test.clj
(ns byte-transforms-test
  (:require
    [criterium.core :as c]
    [clojure.test :refer :all]
    [byte-streams :as bs]
    [byte-transforms :as bt]))

;;;

(defonce warmup-data (bs/to-byte-array (apply str (repeat 1e6 "a"))))
(defonce world-facts (bs/to-byte-array (slurp "test/data/world192.txt")))

(defn benchmark-hash-fn [hash-fn ^bytes data]
  (let [len (alength data)
        now #(System/currentTimeMillis)
        start (now)
        iterations (loop [cnt 0]
                     (if (< (- (now) start) 1000)
                       (do
                         (bt/hash data hash-fn)
                         (recur (inc cnt)))
                       cnt))
        end (now)]
    (float
      (/
        (* len iterations 1000 (Math/pow 2 -20))
        (- end start)))))

(deftest ^:benchmark benchmark-hash-functions
  (println "\nhash function throughput:")
  (doseq [h (sort (bt/available-hash-functions))]
    ;; warmup
    (bt/hash warmup-data h)
    (let [throughput (benchmark-hash-fn h world-facts)]
      (println (format "%12s: %.2f MB/s" h throughput)))))

;;;

(deftest test-compression-roundtrips
  (doseq [c (sort (bt/available-compressors))]
    (testing c
      (is (bt/hash= world-facts
            (-> world-facts
              (bt/compress c)
              (bt/decompress c)
              bs/to-byte-array)))
      (is (bt/hash= world-facts
            (-> world-facts
              bs/to-input-stream
              (bt/compress c)
              (bt/decompress c)
              bs/to-byte-array))))))

(defn benchmark-compression-fn [algorithm ^bytes data]
  (let [len (alength data)
        now #(System/currentTimeMillis)
        measure (fn [f]
                  (let [start (now)]
                    (loop [cnt 0]
                      (if (< (- (now) start) 2000)
                        (do (f) (recur (inc cnt)))
                        (let [end (now)]
                          (float
                            (/
                              (* len cnt 1000 (Math/pow 2 -20))
                              (- end start))))))))
        compressed-data (-> data (bt/compress algorithm) bs/to-byte-array)]
    [(measure #(-> data (bt/compress algorithm) (bt/decompress algorithm) bs/to-byte-array))
     (measure #(-> data (bt/compress algorithm) bs/to-byte-array))
     (measure #(-> compressed-data (bt/decompress algorithm) bs/to-byte-array))]))

(defn measure-compression-fn [algorithm ^bytes data]
  (float
    (/
      (alength data)
      (-> data (bt/compress algorithm) bs/to-byte-array alength))))

(deftest ^:benchmark benchmark-compression-algorithms
  (println "\ncompression throughput:")
  (doseq [c (sort (bt/available-compressors))]
    ;; warmup
    (bt/compress warmup-data c)
    (let [[roundtrip compress decompress] (benchmark-compression-fn c world-facts)]
      (println (format "%12s: roundtrip %.2f MB/s, compress %.2f MB/s, decompress %.2f MB/s"
                 c
                 roundtrip
                 compress
                 decompress))))
  (println "\ncompression factor:")
    (doseq [c (sort (bt/available-compressors))]
      (println (format "%12s: %.2fx" c (measure-compression-fn c world-facts)))))

;;;

(deftest test-encoding-roundtrips
  (doseq [e (sort (bt/available-encoders))]
    (testing e
      (is (bt/hash= world-facts
            (-> world-facts
              (bt/encode e)
              (bt/decode e)
              bs/to-byte-array))))))

(defn benchmark-encoding-fn [algorithm ^bytes data]
  (let [len (alength data)
        now #(System/currentTimeMillis)
        start (now)
        iterations (loop [cnt 0]
                     (if (< (- (now) start) 1000)
                       (do
                         (-> data
                           (bt/encode algorithm)
                           (bt/encode algorithm))
                         (recur (inc cnt)))
                       cnt))
        end (now)]
    (float
      (/
        (* len iterations 1000 (Math/pow 2 -20))
        (- end start)))))

(deftest ^:benchmark benchmark-encodings
  (println "\nencoding roundtrip throughput:")
  (doseq [c (sort (bt/available-encoders))]
    ;; warmup
    (bt/encode warmup-data c)
    (let [throughput (benchmark-encoding-fn c world-facts)]
      (println (format "%12s: %.2f MB/s" c throughput)))))

;;;

(deftest ^:benchmark benchmark-hash-function-latency
  (let [s (apply str (repeat 32 "a"))
        b (bs/to-byte-array s)]
    (println "bench murmur64 32-byte string")
    (c/quick-bench
      (bt/hash s :murmur64))
    (println "bench murmur64 32-byte byte-array")
    (c/quick-bench
      (bt/hash b :murmur64))))


test/byte_transforms_simple_check.clj
(ns byte-transforms-simple-check
  (:require
    [clojure.test :refer :all]
    [byte-streams :as bs]
    [byte-transforms :as bt]
    [clojure.test.check.clojure-test :as ct :refer (defspec)]
    [clojure.test.check
     [generators :as gen]
     [properties :as prop]])
  (:import (java.util Arrays)))

(def compression-type (gen/elements (bt/available-compressors)))

(def concat-compression-type (gen/elements [:gzip :bzip2 :snappy]))

(def not-empty-byte-array (gen/such-that #(pos? (count %)) gen/bytes))

(defn roundtrip-equiv
  [b comp-type]
  (Arrays/equals
    ^bytes b
    (-> b
        (bt/compress comp-type)
        (bt/decompress comp-type)
        bs/to-byte-array)))

(defn concat-roundtrip-equiv
  [b chunk-size comp-type]
  (Arrays/equals
    ^bytes b
    (->> (bs/to-byte-buffers b {:chunk-size chunk-size})
         (map #(bt/compress % comp-type))
         (#(bt/decompress % comp-type))
         bs/to-byte-array)))

(def roundtrip-property
  "Forall byte-arrays `b`, and compression types `comp-type`,
  compressing and then decompressing should be equal to the
  original byte-array"
  (prop/for-all [b not-empty-byte-array, comp-type compression-type]
    (roundtrip-equiv b comp-type)))

(defspec roundtrip-compressors 1e4 roundtrip-property)

(def concat-roundtrip-property
  (prop/for-all
    [b not-empty-byte-array
     chunk-size (gen/such-that pos? gen/pos-int)
     comp-type concat-compression-type]
    (concat-roundtrip-equiv b chunk-size comp-type)))

(defspec concat-roundtrip-compressors 1e4 concat-roundtrip-property)


project.clj
(defproject org.clj-commons/byte-transforms (or (System/getenv "PROJECT_VERSION") "0.2.1")
  :description "Methods for hashing, compressing, and encoding bytes."
  :license {:name "Apache License 2.0"
            :url "http://www.apache.org/licenses/LICENSE-2.0.html"}
  :deploy-repositories [["clojars" {:url "https://repo.clojars.org"
                                    :username :env/clojars_username
                                    :password :env/clojars_password
                                    :sign-releases true}]]
  :dependencies [[byte-streams "0.2.0"]
                 [org.xerial.snappy/snappy-java "1.1.8.4"]
                 [commons-codec/commons-codec "1.10"]
                 [net.jpountz.lz4/lz4 "1.3"]
                 [org.apache.commons/commons-compress "1.20"]]
  :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.1"]
                                  [criterium "0.4.6"]
                                  [org.clojure/test.check "1.1.0"]
                                  [codox-md "0.2.0" :exclusions [org.clojure/clojure]]]}
             :ci {:javac-options ["-target" "1.8" "-source" "1.8"]
                  :dependencies [[org.clojure/clojure "1.10.1"]
                                 [criterium "0.4.6"]
                                 [org.clojure/test.check "1.1.0"]]}}
  :plugins [[codox "0.6.4"]]
  :codox {:writer codox-md.writer/write-docs
          :include [byte-transforms]}
  :java-source-paths ["src"]
  :global-vars {*warn-on-reflection* true}
  :jvm-opts ^:replace ["-server"]
  :test-selectors {:default (complement :benchmark)
                   :benchmark :benchmark})


src/byte_transforms/CRC64.java
/*
 * CRC64
 *
 * Author: Lasse Collin <lasse.collin@tukaani.org>
 *
 * This file has been put into the public domain.
 * You can do whatever you want with this file.
 */

// this file has been slightly modified, but is otherwise
// as originally written by Lasse Collin
package byte_transforms;

public class CRC64 {

    private static final long poly = 0xC96C5795D7870F42L;
    private static final long crcTable[] = new long[256];

    private long crc = -1;

    static {
        for (int b = 0; b < crcTable.length; ++b) {
                long r = b;
                for (int i = 0; i < 8; ++i) {
                        if ((r & 1) == 1)
                                r = (r >>> 1) ^ poly;
                        else
                                r >>>= 1;
                }

                crcTable[b] = r;
        }
    }

    public CRC64() {
    }

    public void update(byte b) {
        crc = crcTable[(b ^ (int)crc) & 0xFF] ^ (crc >>> 8);
    }

    public void update(byte[] buf) {
        update(buf, 0, buf.length);
    }

    public void update(byte[] buf, int off, int len) {
        int end = off + len;

        while (off < end)
            crc = crcTable[(buf[off++] ^ (int)crc) & 0xFF] ^ (crc >>> 8);
    }

    public long getValue() {
        return ~crc;
    }
}


src/byte_transforms/CassandraMurmurHash.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package byte_transforms;

import java.nio.ByteBuffer;

/**
 * This is a very fast, non-cryptographic hash suitable for general hash-based
 * lookup. See http://murmurhash.googlepages.com/ for more details.
 *
 * hash32() and hash64() are MurmurHash 2.0.
 * hash3_x64_128() is MurmurHash 3.0.
 *
 * <p>
 * The C version of MurmurHash 2.0 found at that site was ported to Java by
 * Andrzej Bialecki (ab at getopt org).
 * </p>
 */
public class CassandraMurmurHash
{
    public static int hash32(ByteBuffer data, int offset, int length, int seed)
    {
        int m = 0x5bd1e995;
        int r = 24;

        int h = seed ^ length;

        int len_4 = length >> 2;

        for (int i = 0; i < len_4; i++)
        {
            int i_4 = i << 2;
            int k = data.get(offset + i_4 + 3);
            k = k << 8;
            k = k | (data.get(offset + i_4 + 2) & 0xff);
            k = k << 8;
            k = k | (data.get(offset + i_4 + 1) & 0xff);
            k = k << 8;
            k = k | (data.get(offset + i_4 + 0) & 0xff);
            k *= m;
            k ^= k >>> r;
            k *= m;
            h *= m;
            h ^= k;
        }

        // avoid calculating modulo
        int len_m = len_4 << 2;
        int left = length - len_m;

        if (left != 0)
        {
            if (left >= 3)
            {
                h ^= (int) data.get(offset + length - 3) << 16;
            }
            if (left >= 2)
            {
                h ^= (int) data.get(offset + length - 2) << 8;
            }
            if (left >= 1)
            {
                h ^= (int) data.get(offset + length - 1);
            }

            h *= m;
        }

        h ^= h >>> 13;
        h *= m;
        h ^= h >>> 15;

        return h;
    }

    public static long hash2_64(ByteBuffer key, int offset, int length, long seed)
    {
        long m64 = 0xc6a4a7935bd1e995L;
        int r64 = 47;

        long h64 = (seed & 0xffffffffL) ^ (m64 * length);

        int lenLongs = length >> 3;

        for (int i = 0; i < lenLongs; ++i)
        {
            int i_8 = i << 3;

            long k64 =  ((long)  key.get(offset+i_8+0) & 0xff)      + (((long) key.get(offset+i_8+1) & 0xff)<<8)  +
                        (((long) key.get(offset+i_8+2) & 0xff)<<16) + (((long) key.get(offset+i_8+3) & 0xff)<<24) +
                        (((long) key.get(offset+i_8+4) & 0xff)<<32) + (((long) key.get(offset+i_8+5) & 0xff)<<40) +
                        (((long) key.get(offset+i_8+6) & 0xff)<<48) + (((long) key.get(offset+i_8+7) & 0xff)<<56);

            k64 *= m64;
            k64 ^= k64 >>> r64;
            k64 *= m64;

            h64 ^= k64;
            h64 *= m64;
        }

        int rem = length & 0x7;

        switch (rem)
        {
        case 0:
            break;
        case 7:
            h64 ^= (long) key.get(offset + length - rem + 6) << 48;
        case 6:
            h64 ^= (long) key.get(offset + length - rem + 5) << 40;
        case 5:
            h64 ^= (long) key.get(offset + length - rem + 4) << 32;
        case 4:
            h64 ^= (long) key.get(offset + length - rem + 3) << 24;
        case 3:
            h64 ^= (long) key.get(offset + length - rem + 2) << 16;
        case 2:
            h64 ^= (long) key.get(offset + length - rem + 1) << 8;
        case 1:
            h64 ^= (long) key.get(offset + length - rem);
            h64 *= m64;
        }

        h64 ^= h64 >>> r64;
        h64 *= m64;
        h64 ^= h64 >>> r64;

        return h64;
    }

    protected static long getblock(ByteBuffer key, int offset, int index)
    {
        int i_8 = index << 3;
        int blockOffset = offset + i_8;
        return ((long) key.get(blockOffset + 0) & 0xff) + (((long) key.get(blockOffset + 1) & 0xff) << 8) +
               (((long) key.get(blockOffset + 2) & 0xff) << 16) + (((long) key.get(blockOffset + 3) & 0xff) << 24) +
               (((long) key.get(blockOffset + 4) & 0xff) << 32) + (((long) key.get(blockOffset + 5) & 0xff) << 40) +
               (((long) key.get(blockOffset + 6) & 0xff) << 48) + (((long) key.get(blockOffset + 7) & 0xff) << 56);
    }

    protected static long rotl64(long v, int n)
    {
        return ((v << n) | (v >>> (64 - n)));
    }

    protected static long fmix(long k)
    {
        k ^= k >>> 33;
        k *= 0xff51afd7ed558ccdL;
        k ^= k >>> 33;
        k *= 0xc4ceb9fe1a85ec53L;
        k ^= k >>> 33;

        return k;
    }

    public static long[] hash3_x64_128(ByteBuffer key, int offset, int length, long seed)
    {
        final int nblocks = length >> 4; // Process as 128-bit blocks.

        long h1 = seed;
        long h2 = seed;

        long c1 = 0x87c37b91114253d5L;
        long c2 = 0x4cf5ad432745937fL;

        //----------
        // body

        for(int i = 0; i < nblocks; i++)
        {
            long k1 = getblock(key, offset, i*2+0);
            long k2 = getblock(key, offset, i*2+1);

            k1 *= c1; k1 = rotl64(k1,31); k1 *= c2; h1 ^= k1;

            h1 = rotl64(h1,27); h1 += h2; h1 = h1*5+0x52dce729;

            k2 *= c2; k2  = rotl64(k2,33); k2 *= c1; h2 ^= k2;

            h2 = rotl64(h2,31); h2 += h1; h2 = h2*5+0x38495ab5;
        }

        //----------
        // tail

        // Advance offset to the unprocessed tail of the data.
        offset += nblocks * 16;

        long k1 = 0;
        long k2 = 0;

        switch(length & 15)
        {
            case 15: k2 ^= ((long) key.get(offset+14)) << 48;
            case 14: k2 ^= ((long) key.get(offset+13)) << 40;
            case 13: k2 ^= ((long) key.get(offset+12)) << 32;
            case 12: k2 ^= ((long) key.get(offset+11)) << 24;
            case 11: k2 ^= ((long) key.get(offset+10)) << 16;
            case 10: k2 ^= ((long) key.get(offset+9)) << 8;
            case  9: k2 ^= ((long) key.get(offset+8)) << 0;
                k2 *= c2; k2  = rotl64(k2,33); k2 *= c1; h2 ^= k2;

            case  8: k1 ^= ((long) key.get(offset+7)) << 56;
            case  7: k1 ^= ((long) key.get(offset+6)) << 48;
            case  6: k1 ^= ((long) key.get(offset+5)) << 40;
            case  5: k1 ^= ((long) key.get(offset+4)) << 32;
            case  4: k1 ^= ((long) key.get(offset+3)) << 24;
            case  3: k1 ^= ((long) key.get(offset+2)) << 16;
            case  2: k1 ^= ((long) key.get(offset+1)) << 8;
            case  1: k1 ^= ((long) key.get(offset));
                k1 *= c1; k1  = rotl64(k1,31); k1 *= c2; h1 ^= k1;
        };

        //----------
        // finalization

        h1 ^= length; h2 ^= length;

        h1 += h2;
        h2 += h1;

        h1 = fmix(h1);
        h2 = fmix(h2);

        h1 += h2;
        h2 += h1;

        return(new long[] {h1, h2});
    }
}


src/byte_transforms.clj
(ns byte-transforms
  (:refer-clojure :exclude (byte-array hash))
  (:require
    [byte-streams :refer (seq-of bytes=) :as bytes]
    [primitive-math :as p])
  (:import
    [byte_transforms
     CassandraMurmurHash
     CRC64]
    [java.util
     UUID]
    [java.lang.reflect
     Array]
    [java.util.zip
     CRC32
     Adler32]
    [java.io
     OutputStream
     InputStream
     PipedInputStream
     PipedOutputStream
     ByteArrayOutputStream]
    [net.jpountz.lz4
     LZ4BlockOutputStream
     LZ4BlockInputStream
     LZ4Factory
     LZ4Compressor]
    [java.security
     MessageDigest]
    [java.nio
     DirectByteBuffer
     ByteBuffer
     ShortBuffer
     IntBuffer
     LongBuffer]
    [org.apache.commons.codec.binary
     Base64]
    [org.apache.commons.compress.compressors.bzip2
     BZip2CompressorInputStream
     BZip2CompressorOutputStream]
    [org.apache.commons.compress.compressors.gzip
     GzipCompressorInputStream
     GzipCompressorOutputStream
     GzipParameters]
    [org.xerial.snappy
     Snappy
     SnappyInputStream
     SnappyOutputStream]))

;;;

(def ^:private hash-functions (atom {}))
(def ^:private compressors (atom {}))
(def ^:private decompressors (atom {}))
(def ^:private encoders (atom {}))
(def ^:private decoders (atom {}))

(defmacro def-hash [hash-name [bytes options] & body]
  `(swap! byte-transforms/hash-functions assoc ~(keyword (name hash-name))
     (fn [~bytes ~(or options (gensym "options"))]
       ~@body)))

(defmacro def-compressor [compressor-name [bytes options] & body]
  `(swap! byte-transforms/compressors assoc ~(keyword (name compressor-name))
     (fn [~bytes ~(or options (gensym "options"))]
       ~@body)))

(defmacro def-decompressor [decompressor-name [bytes options] & body]
  `(swap! byte-transforms/decompressors assoc ~(keyword (name decompressor-name))
     (fn [~bytes ~(or options (gensym "options"))]
       ~@body)))

(defmacro def-encoder [encoder-name [bytes options] & body]
  `(swap! byte-transforms/encoders assoc ~(keyword (name encoder-name))
     (fn [~bytes ~(or options (gensym "options"))]
       ~@body)))

(defmacro def-decoder [decoder-name [bytes options] & body]
  `(swap! byte-transforms/decoders assoc ~(keyword (name decoder-name))
     (fn [~bytes ~(or options (gensym "options"))]
       ~@body)))

;;;

(def ^:private ^:const byte-array (class (clojure.core/byte-array 0)))

(defn available-hash-functions
  "Returns the name of all available hash functions."
  []
  (keys @hash-functions))

(defn hash=
  "Returns true if the two hashes equal each other"
  [a b]
  (if (and
        (instance? byte-array a)
        (instance? byte-array b))
    (bytes= a b)
    (= a b)))

(defn hash->bytes
  "Converts a hash to an array of bytes."
  [x]
  (condp instance? x
    byte-array x
    Integer   (-> (ByteBuffer/allocate 4) (.putInt (int x)) .array)
    Long      (-> (ByteBuffer/allocate 8) (.putLong (long x)) .array)))

(defn hash->shorts
  "Converts a hash to an array of shorts."
  [x]
  (let [^bytes bytes (hash->bytes x)
        ary (short-array (p/>> (alength bytes) 1))]
    (-> bytes ByteBuffer/wrap .asShortBuffer (.get ary))
    ary))

(defn hash->ints
  "Converts a hash to an array of integers."
  [x]
  (let [^bytes bytes (hash->bytes x)
        ary (int-array (p/>> (alength bytes) 2))]
    (-> bytes ByteBuffer/wrap .asIntBuffer (.get ary))
    ary))

(defn hash->longs
  "Converts a hash to an array of longs."
  [x]
  (let [^bytes bytes (hash->bytes x)
        ary (long-array (p/>> (alength bytes) 3))]
    (-> bytes ByteBuffer/wrap .asLongBuffer (.get ary))
    ary))

(defn hash->uuid
  "Converts a 128-bit hash to a string representation of a UUID."
  [x]
  (let [^longs ary (hash->longs x)]
    (when-not (== 2 (Array/getLength ary))
      (throw (IllegalArgumentException. "Expected 128 bit input.")))
    (str (UUID. (aget ary 0) (aget ary 1)))))

;; CRC32 hash
(def-hash crc32
  [x options]
  (let [crc (CRC32.)]
    (when-let [seed (get options :seed)]
      (.update crc (byte seed)))
    (doseq [^bytes ary (bytes/to-byte-arrays x options)]
      (.update crc ary))
    (.getValue crc)))

;; CRC64 hash
(def-hash crc64
  [x options]
  (let [crc (CRC64.)]
    (when-let [seed (get options :seed)]
      (.update crc (byte seed)))
    (doseq [^bytes ary (bytes/to-byte-arrays x options)]
      (.update crc ary))
    (.getValue crc)))

;; Adler32 hash
(def-hash adler32
  [x options]
  (let [adler (Adler32.)]
    (when-let [seed (get options :seed)]
      (.update adler (byte seed)))
    (doseq [^bytes ary (bytes/to-byte-arrays x options)]
      (.update adler ary))
    (.getValue adler)))

(defn- hash-digest [^MessageDigest digest bufs options]
  (when-let [seed (get options :seed)]
    (.update digest (byte seed)))
  (doseq [^ByteBuffer buf bufs]
    (.update digest buf))
  (.digest digest))

(defmacro ^:private def-digests [& names]
  `(do
     ~@(map
         (fn [[digest-name name]]
           `(def-hash ~name
              [x# options#]
              (hash-digest
                (MessageDigest/getInstance ~(str digest-name))
                (bytes/to-byte-buffers x# options#)
                options#)))
         names)))

;; all standard message digest hashes
(def-digests
  [md2 md2]
  [md5 md5]
  [sha-1 sha1]
  [sha-256 sha256]
  [sha-384 sha384]
  [sha-512 sha512])

;; murmur32 hash
(def-hash murmur32
  [x options]
  (let [buf (bytes/to-byte-buffer x options)]
    (CassandraMurmurHash/hash32
      buf
      0
      (.remaining ^ByteBuffer buf)
      (get options :seed 0))))

;; murmur64 hash
(def-hash murmur64
  [x options]
  (let [buf (bytes/to-byte-buffer x options)]
    (CassandraMurmurHash/hash2_64
      buf
      0
      (.remaining ^ByteBuffer buf)
      (get options :seed 0))))

;; murmur128 hash
(def-hash murmur128
  [x options]
  (let [buf (bytes/to-byte-buffer x options)
        ^longs ls (CassandraMurmurHash/hash3_x64_128
                    buf
                    0
                    (.remaining ^ByteBuffer buf)
                    (get options :seed 0))]
    (-> (ByteBuffer/allocate 16)
      (.putLong (aget ls 0))
      (.putLong (aget ls 1))
      .array)))

(let [murmur32 (get @hash-functions :murmur32)
      murmur64 (get @hash-functions :murmur64)
      murmur128 (get @hash-functions :murmur128)
      crc64 (get @hash-functions :crc64)]
  (defn hash
    "Takes a byte stream, and returns a value representing its hash, which will be an integer if
   the hash is 32 or 64-bit, or a byte array otherwise.  By default, this will use the murmur64
   hash."
    ([bytes]
       (hash bytes :murmur64 nil))
    ([bytes function]
       (hash bytes function nil))
    ([bytes function options]
       (case function
         :murmur32 (murmur32 bytes options)
         :murmur64 (murmur64 bytes options)
         :murmur128 (murmur128 bytes options)
         :crc64 (crc64 bytes options)
         (if-let [f (@hash-functions (keyword function))]
           (f bytes options)
           (throw
             (IllegalArgumentException.
               (str "Don't recognize hash function '" (name function) "'"))))))))

;;;

(defn available-compressors []
  (keys @compressors))

(defn compress
  "Takes a byte stream, and returns a compressed version.  By default, this will use Snappy."
  ([x]
     (compress x :snappy))
  ([x algorithm]
     (compress x algorithm nil))
  ([x algorithm options]
     (if-let [f (@compressors (keyword algorithm))]
       (f x options)
       (throw
         (IllegalArgumentException.
           (str "Don't recognize compressor '" (name algorithm) "'"))))))

(defn decompress
  "Takes a compressed byte stream, and return an uncompressed version.  By default, this will use Snappy."
  ([x]
     (decompress x :snappy))
  ([x algorithm]
     (decompress x algorithm nil))
  ([x algorithm options]
     (if-let [f (@decompressors (keyword algorithm))]
       (f x options)
       (throw
         (IllegalArgumentException.
           (str "Don't recognize decompressor '" (name algorithm) "'"))))))

(defn- in->wrapped-out->in
  [^InputStream stream output-wrapper options]
  (let [chunk-size (get options :chunk-size 65536)
        out (PipedOutputStream.)
        in (PipedInputStream. out chunk-size)
        ^OutputStream compressor (output-wrapper out)]
    (future
      (try
        (let [ary (clojure.core/byte-array chunk-size)]
          (loop []
            (let [n (.read stream ary)]
              (when (pos? n)
                (.write compressor ary 0 n)
                (recur)))))
        (finally
          (.close compressor)
          (.close out))))

    in))

(defn bytes->wrapped-out->bytes
  [bytes output-wrapper options]
  (if (or (instance? ByteBuffer bytes)
        (instance? byte-array bytes))
    (let [^ByteBuffer buf (bytes/to-byte-buffer bytes)
          out (ByteArrayOutputStream.)
          ^OutputStream compressor (output-wrapper out)
          ^bytes ary (clojure.core/byte-array 1024)]
      (loop []
        (let [remaining (.remaining buf)]
          (when (pos? remaining)
            (let [len (p/min remaining 1024)]
              (.get buf ary 0 len)
              (.write compressor ary 0 len)
              (recur)))))
      (.close compressor)
      (.close out)
      (.toByteArray out))
    (in->wrapped-out->in
      (bytes/to-input-stream bytes options)
      output-wrapper
      options)))

(def-compressor gzip
  [x {:keys [compression-level] :as options}]
  (bytes->wrapped-out->bytes
    x
    (if compression-level
      #(GzipCompressorOutputStream. %
         (doto (GzipParameters.)
           (.setCompressionLevel compression-level)))
      #(GzipCompressorOutputStream. %))
    options))

(def-decompressor gzip
  [x options]
  (GzipCompressorInputStream. (bytes/to-input-stream x options) true))

(def-compressor snappy
  [x options]
  (cond
    (instance? byte-array x)
    (Snappy/compress ^bytes x)

    (<= (bytes/conversion-cost x byte-array) (bytes/conversion-cost x (seq-of byte-array)))
    (Snappy/compress (bytes/to-byte-array x options))

    :else
    (map
      #(Snappy/compress ^bytes %)
      (bytes/to-byte-arrays x (update-in options [:chunk-size] #(or % 32278))))))

(def-decompressor snappy
  [x options]
  (cond
    (instance? byte-array x)
    (Snappy/uncompress x)

    (and (not (seq? x))
      (<= (bytes/conversion-cost x byte-array) (bytes/conversion-cost x (seq-of byte-array))))
    (Snappy/uncompress (bytes/to-byte-array x options))

    :else
    (map
      #(Snappy/uncompress ^bytes %)
      (bytes/to-byte-arrays x (update-in options [:chunk-size] #(or % 32278))))))

(def-compressor bzip2
  [x options]
  (bytes->wrapped-out->bytes
    x
    #(BZip2CompressorOutputStream. %)
    options))

(def-decompressor bzip2
  [x options]
  (BZip2CompressorInputStream. (bytes/to-input-stream x options) true))

(def-compressor lz4
  [x {:keys [safe? fastest? chunk-size]
      :or {safe? false, fastest? false, chunk-size 1e5}
      :as options}]
  (bytes->wrapped-out->bytes
    x
    #(LZ4BlockOutputStream. %
       chunk-size
       (let [^LZ4Factory factory (if safe?
                                   (LZ4Factory/safeInstance)
                                   (LZ4Factory/fastestInstance))]
         (if fastest?
           (.fastCompressor factory)
           (.highCompressor factory))))
    options))

(def-decompressor lz4
  [x options]
  (LZ4BlockInputStream. (bytes/to-input-stream x options)))

;;;

(defn available-encoders
  "Returns a list of all available encodings."
  []
  (keys @encoders))

(defn encode
  "Takes a byte stream, and returns an encoded version."
  ([x encoding]
     (encode x encoding nil))
  ([x encoding options]
     (if-let [f (@encoders (keyword encoding))]
       (f x options)
       (throw
         (IllegalArgumentException.
           (str "Don't recognize encoding '" encoding "'"))))))

(defn decode
    "Takes an encoded byte stream, and returns a decoded version."
  ([x encoding]
     (decode x encoding nil))
  ([x encoding options]
     (if-let [f (@decoders (keyword encoding))]
       (f x options)
       (throw
         (IllegalArgumentException.
           (str "Don't recognize decompressor '" (name encoding) "'"))))))

(def-encoder base64
  [x {:keys [url-safe? line-length line-separator]
      :or {url-safe? true}
      :as options}]
  (if line-length
    (let [encoder (Base64.
                    (or line-length 76)
                    (or line-separator (.getBytes "\r\n"))
                    url-safe?)]
      (.encode encoder (bytes/to-byte-array x options)))
    (Base64/encodeBase64
      (bytes/to-byte-array x options)
      false
      url-safe?)))

(def-decoder base64
  [x options]
  (Base64/decodeBase64 (bytes/to-byte-array x options)))
